### PR TITLE
core/manager: pass soft-reboot count to generators

### PR DESCRIPTION
**Pull request by YHNdnzj**

### CODE REVIEW

**Review by yuwata — COMMENTED** [review]
Please add test cases.

File: src/shared/exec-util.c [hunk 1]
```diff
@@ -438,6 +438,12 @@ static int gather_environment_consume(int fd, void *arg) {
         return r;
 }
 
+const gather_stdout_callback_t gather_environment[] = {
```
Comment (yuwata) [inline src/shared/exec-util.c]: I do not see any reason to move this.

File: src/shared/exec-util.h [hunk 2]
```diff
@@ -14,24 +14,14 @@ enum {
...
-        EXEC_DIR_NONE                 = 0,      /* No execdir flags */
+typedef enum ExecDirFlags {
         EXEC_DIR_PARALLEL             = 1 << 0, /* Execute scripts in parallel, if possible */
         EXEC_DIR_IGNORE_ERRORS        = 1 << 1, /* Ignore non-zero exit status of scripts */
         EXEC_DIR_SET_SYSTEMD_EXEC_PID = 1 << 2, /* Set $SYSTEMD_EXEC_PID environment variable */
         EXEC_DIR_SKIP_REMAINING       = 1 << 3, /* Ignore remaining executions when one exit with 77. */
         EXEC_DIR_WARN_WORLD_WRITABLE  = 1 << 4, /* Warn if world writable files are found */
 } ExecDirFlags;
 
-typedef enum ExecCommandFlags {
```
Comment (yuwata) [inline src/shared/exec-util.h:26]: Not a strict rule, but we typically declare types earlier in header. So, I'd like to keep this here.

File: src/fstab-generator/fstab-generator.c [hunk 3]
```diff
@@ -1459,20 +1459,29 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
 
         } else if (streq(key, "root")) {
 
+                if (generator_soft_rebooted())
```
Comment (yuwata) [inline src/fstab-generator/fstab-generator.c]: IIRC, root= and friends are used only used in initrd.
But, after soft-reboot system will not enter initrd, right??
In what kind of situation is this necessary??

At least, I think it is better to move the check to add_sysroot_mount() and add_sysroot_usr_mount_fallback().

File: src/fstab-generator/fstab-generator.c [hunk 4]
```diff
@@ -1510,20 +1531,25 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
...
-        } else if (streq(key, "rw") && !value)
+        } else if (streq(key, "rw") && !value && !generator_soft_rebooted())
                 arg_root_rw = true;
-        else if (streq(key, "ro") && !value)
+        else if (streq(key, "ro") && !value && !generator_soft_rebooted())
                 arg_root_rw = false;
         else if (streq(key, "systemd.volatile")) {
-                VolatileMode m;
+
+                if (generator_soft_rebooted())
```
Comment (yuwata) [inline src/fstab-generator/fstab-generator.c]: This means, even if booted with `systemd.volatile`, disk will be writable after soft-reboot. I think it is dangerous. No?

File: src/shared/exec-util.c (hunk 1 above)
Comment (poettering) [reply src/shared/exec-util.c]: if you touch this I'd probably add `[_STDOUT_CONSUME_MAX]` here, to fixate the size. doesn't really change anything, but given that the size is part of the API here, it's probably worth declaring it explicitly.

File: src/test/test-exec-util.c [hunk 5]
```diff
@@ -402,7 +402,7 @@ TEST(error_catching) {
         if (access(name, X_OK) < 0 && ERRNO_IS_PRIVILEGE(errno))
                 return;
 
-        r = execute_directories(dirs, DEFAULT_TIMEOUT_USEC, NULL, NULL, NULL, NULL, EXEC_DIR_NONE);
+        r = execute_directories(dirs, DEFAULT_TIMEOUT_USEC, NULL, NULL, NULL, NULL, 0);
```
Comment (poettering) [inline src/test/test-exec-util.c]: maybe insert `/* flags= */` comment though?

(and maybe for the 4 NULL args too?)

File: man/systemd.generator.xml [hunk 6]
```diff
@@ -163,6 +163,15 @@
         <xi:include href="version-info.xml" xpointer="v251"/></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>$SYSTEMD_SOFT_REBOOTS_COUNT</varname></term>
+
+        <listitem><para>If the system has soft-rebooted, this variable is set to the count of soft-reboots.
+        This environment variable is only set for system generators.</para>
```
Comment (poettering) [inline man/systemd.generator.xml:170]: maybe say that this numeric value of this variable will be > 0 always.

**Comment by poettering** [comment]
Hmm, we run all generators on daemon-reload and so on, and I think it's *fine* and even *good* if they generate units for the root fs and stuff then. I don't follow why you say that this would be bad in soft-reboot case?

Genreating should generally be safe – as long as they are not actually run again (and they shouldn't be, as the root fs is already mounted by then).

Hence, can you elaborate what the last commit is really about? (the other ones look ok to me)

File: src/shared/exec-util.c (hunk 1 above)
Comment (YHNdnzj) [reply src/shared/exec-util.c]: > I do not see any reason to move this.

This is part of `execute_directories` API above, while functions below are about `ExecCommandFlags`.

File: src/shared/exec-util.h (hunk 2 above)
Comment (YHNdnzj) [reply src/shared/exec-util.h:26]: See above. The two enums are completely irrelevant, hence I prefer to place them closer to corresponding functions.

We do this also in e.g. basic/fileio.h

File: man/systemd.generator.xml (hunk 6 above)
Comment (YHNdnzj) [reply man/systemd.generator.xml:170]: hmm, I see this as an implementation detail and doesn't really worth tightening the documentation for? Generator impls shall still parse the numeric value before use.

**Comment by YHNdnzj** [comment]
OK, dropped the last commit, except for the hibernate-resume part.

File: src/core/manager.c [hunk 7]
```diff
@@ -4007,40 +4007,43 @@ void manager_send_reloading(Manager *m) {
...
         paths = env_generator_binary_paths(m->runtime_scope);
         if (!paths)
                 return log_oom();
 
-        if (!generator_path_any((const char* const*) paths))
+        if (!generator_path_any(paths))
                 return 0;
 
+        char **tmp = NULL; /* this is only used in the forked process, no cleanup here */
+        void *args[] = {
```
Comment (yuwata) [inline src/core/manager.c]: Here also `[]` -> `[_STDOUT_CONSUME_MAX]` ?

**Review by yuwata — APPROVED** [review]
LGTM. One minor comment.

**Comment by yuwata** [comment]
CI failures caused by recent imaged change, unrelated.

**Pull request merged**